### PR TITLE
feat(event-handler): Add support for HTTP APIs (API Gateway v2)

### LINKED ISSUE #4713: Feature request: Add support for HTTP APIs (API Gateway v2) in Event Handler
### Use case

Currently event handler only supports API Gateway Rest APIs. We should add support for HTTP APIs.

### Solution/User Experience

The event source should be transparent to users, they should be able to define handlers that return any valid API Gateway response we support (JSON object, `APIGatewayProxyEvent`, `APIGatewayProxyEventV2`, Web `Response` object) and event handler will automatically detect the input event and convert to the appropriate response.

```ts
const app = new Router();

// this returns APIGatewayProxyEventV2
app.get('/test', () => ({
    statusCode: 200,
    body: JSON.stringify({ message: 'success' }),
    headers: { 'content-type': 'application/json' },
    cookies: ['cookie1', 'cookie2'],
}));

const v1Result = await app.resolve(aV1Event, context);

/**
v1Result = {
    "statusCode": 200,
    "headers": {
        "content-type": "application/json",
    },
    "body": "{\"message\":\"success\"}",
    "isBase64Encoded": false,
    "multiValueHeaders": {
        "set-cookie": [
            "cookie1",
            "cookie2"
        ]
    }
}
 */

const v2Result = await app.resolve(aV2Event, context);

/**
v2Result = {
    "statusCode": 200,
    "headers": {
        "content-type": "application/json",
    },
    "body": "{\"message\":\"success\"}",
    "isBase64Encoded": false,
    "cookies": [
        "cookie1",
        "cookie2"
    ]
}
 **/
    
```

### Alternative solutions

```Markdown
Alternative solution would be for customers to have specific resolvers for specific event sources:


const httpApiRouter = new HttpApiRouter()
// will only work with HttpApi events
```

### Acknowledgment

- [x] This feature request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.aws.amazon.com/powertools/typescript/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

### PR DESCRIPTION
**Pull request by svozza**
## Summary
This PR adds support for HTTP API (API Gateway v2) to the event handler utility. It also adds the framework for adding other event sources like function URLs, ALB and VPC lattice.

Edit: after some testing, I've realised that as function URLs and HTTP APIs share the same event type, which means function URLs also work.

### Changes
- The public `resolve` method now has overloads to support both `APIGatewayProxyEvent` and `APIGatewayProxyEventV2` events. New overloads can be added for new event types. I attempted to do this with conditional types but was unable to make it it work, furthermore, the type signatures were much harder to read than simpler overrides.
- The `#resolve` method in the Router class now returns the entire request context object. We use `reqCtx.res` now as the canonical representation of the response to be returned. As the request is now always represented as a web `Response` object, converting the response required for different event sources is simpler. We also save the type of response required in the request context using the `ResponseType` type.
- The `handlerResultToProxyResult` function has been removed now that all response are converted to Web `Response` objects.
- I have removed the ad hoc base64 encoding functionality. The change to the `#resolve` method will allow us to handle this is a more principled way by using the request context to indicate whether a response should be encoded or not. I plan to do this work when implementing https://github.com/aws-powertools/powertools-lambda-typescript/issues/4514.
- A new type guard to identify `APIGatewayProxyEventV2` objects has been added.
- A function to convert `APIGatewayProxyEventV2` events to Web `Request` objects has been added.
- A function to convert Web `Respsonse` objects to `APIGatewayProxyReultV2` objects has been added.

# Testing
- Most of the tests now run using both v1 events and v2 events. The pattern established is easily extended to use additional event types. One suite I have not done so is the `middleware` tests. These use the exact same mechanisms as the other tests so I have decided to only add some additional  `v2` specific tests. I am open to persusion as to whether these should follow the other tests.
- I have used the same lambda function to integrate with an API Gateway REST API and an HTTP API and verified using `curl` that requests will be fulfilled by the same code with no changes required.
- Tested a lambda function URL with streaming enabled and it also worked.

**Issue number:** closes #4713 

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: packages/event-handler/tests/unit/rest/Router/basic-routing.test.ts [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
-  InternalServerError,
   Router,
 } from '../../../../src/rest/index.js';
 import type { HttpMethod, RouteHandler } from '../../../../src/types/rest.js';
-import { createTestEvent } from '../helpers.js';
+import { createTestEvent, createTestEventV2 } from '../helpers.js';
 
-describe('Class: Router - Basic Routing', () => {
+describe.each([
+  { version: 'V1', createEvent: createTestEvent },
```
Comment (svozza) [inline packages/event-handler/tests/unit/rest/Router/basic-routing.test.ts:13]: As mentioned, these tests can be extended just by adding a new test event creation function to the `describe.each` block.

File: packages/event-handler/src/rest/converters.ts [hunk 2]
```diff
@@ -285,26 +447,10 @@ const bodyToNodeStream = (body: ExtendedAPIGatewayProxyResultBody) => {
...
 };
 
-/**
- * Converts a Node.js Readable stream to a base64 encoded string.
- * Handles both Buffer and string chunks by converting all to Buffers.
- *
- * @param stream - The Node.js Readable stream to convert
- * @returns A Promise that resolves to a base64 encoded string
- */
-async function nodeStreamToBase64(stream: Readable) {
```
Comment (svozza) [inline packages/event-handler/src/rest/converters.ts:295]: We will need to add this back when we implement #4514.

File: packages/event-handler/src/rest/Router.ts [hunk 3]
```diff
@@ -202,38 +209,45 @@ class Router {
...
+      req = proxyEventToWebRequest(event);
+    } catch (err) {
+      if (err instanceof InvalidHttpMethodError) {
+        this.logger.error(err);
+        // We can't throw a MethodNotAllowedError outside the try block as it
+        // will be converted to an internal server error by the API Gateway runtime
+        return {
+          event,
+          context,
+          req: new Request('https://invalid'),
```
Comment (svozza) [inline packages/event-handler/src/rest/Router.ts:236]: I'm not fond of creating this dummy `Request` object but the alternative is to make `req` optional, which will make the ergonomics of writing middleware much worse: users would need to do a `null` check any time they want to interact with a request that will always be there apart from this particular corner case.

File: packages/event-handler/tests/unit/rest/Router/error-handling.test.ts [hunk 4]
```diff
@@ -567,3 +588,32 @@ describe('Class: Router - Error Handling', () => {
...
 });
+describe('Class: Router - proxyEventToWebRequest Error Handling', () => {
+  beforeEach(() => {
+    vi.resetModules();
+  });
+
+  it('re-throws non-InvalidHttpMethodError from proxyEventToWebRequest', async () => {
+    // Prepare
+    vi.doMock('../../../../src/rest/converters.js', async () => {
+      const actual = await vi.importActual<
```
Comment (svozza) [inline packages/event-handler/tests/unit/rest/Router/error-handling.test.ts:599]: I very much dislike tests like this. The reason we have to do this is because the type guards before this code path are so strict that the error code is virtually unreachable. I have to mock the `proxyEventToWebRequest` function to throw an exception that it will almost certainly never throw in real use.

File: packages/event-handler/src/rest/converters.ts [hunk 5]
```diff
@@ -81,8 +89,68 @@ const proxyEventToWebRequest = (event: APIGatewayProxyEvent): Request => {
...
+  }
+
+  if (Array.isArray(event.cookies)) {
+    headers.set('Cookie', event.cookies.join('; '));
+  }
+
+  const hostname = headers.get('Host') ?? domainName;
+  const protocol = headers.get('X-Forwarded-Proto') ?? 'https';
+
+  const url = rawQueryString
```
Comment (sdangol) [inline packages/event-handler/src/rest/converters.ts:117]: Should we explicitly check for an empty string?

File: packages/event-handler/tests/unit/rest/Router/middleware.test.ts [hunk 6]
```diff
@@ -595,8 +596,8 @@ describe('Class: Router - Middleware', () => {
 
       // Assess
       expect(result.statusCode).toBe(200);
-      expect(result.isBase64Encoded).toBe(true);
-      expect(result.body).toBe(Buffer.from(testData).toString('base64'));
+      expect(result.isBase64Encoded).toBe(false);
```
Comment (sdangol) [inline packages/event-handler/tests/unit/rest/Router/middleware.test.ts:599]: Was this because of removing `nodeStreamToBase64`? Does this mean compressed response aren't supported now?

Comment (svozza) [reply packages/event-handler/tests/unit/rest/Router/middleware.test.ts:599]: No, this was the ad hoc compression I mentioned in the summary. We were automatically compressing a response if you returned a stream but it only worked for a middleware that returned a response. The compression response is handled in `webResponseToProxyResultV1` and `webResponseToProxyResultV2`, so is unaffected. I want to ensure the logic for compression is more centralised and this is the first phase of that.

Comment (sdangol) [reply packages/event-handler/tests/unit/rest/Router/middleware.test.ts:599]: Ok, got it. Thanks for explaining

File: packages/event-handler/src/rest/converters.ts (hunk 5 above)
Comment (svozza) [reply packages/event-handler/src/rest/converters.ts:117]: I think this is one of the few times we actually want falsy behaviour, we want to skip on `null`, `undefined` and empty string here. I think the type in `aws-lambda` is wrong here, it says this field can't be `undefined` but if you look at the Rust SDK it's marked as optional.

Comment (sdangol) [reply packages/event-handler/src/rest/converters.ts:117]: Ok, makes sense

Comment (svozza) [reply packages/event-handler/src/rest/converters.ts:117]: Actually nvm the Rust bit it has lots of things marked optional such as even `rawPath` so I think they've been over-zealous in their definition. But I don't think it does any harm to be defensive here.

**Comment by svozza** [comment]
Nice!

**Comment by svozza** [comment]
@swopnildangol I made a change to how we calculate what response to return. Rather than passing the event around and using type guards, I save the response type once in the request context with a type called `ResponseType` that can only have the value `v1` or `v2`. This allows us to remove a bunch of overloads in the conversion functions and just use conditional types.

**Comment by sdangol** [comment]
> @swopnildangol I made a change to how we calculate what response to return. Rather than passing the event around and using type guards, I save the response type once in the request context with a type called `ResponseType` that can only have the value `v1` or `v2`. This allows us to remove a bunch of overloads in the conversion functions and just use conditional types.

So, with this, as we add on support for more handlers, we'd be adding options to the ResponseType? Would it make more sense to have maybe ApiGatewayV1 or ApiGatewayV2 rather than just v1 and v2?

**Comment by svozza** [comment]
Good point, I've changed it.

File: .github/workflows/make-release.yml [hunk 7]
```diff
@@ -5,7 +5,7 @@ name: Make Release
 # === Automated activities ===
 # 1. [Quality check] run unit tests, linting, examples, layer, doc snippets
 # 2. [Release] publish all packages to npmjs.org using the latest git commit, ensure provenance with NPM_CONFIG_PROVENANCE=true
-# 3. [Create tag] create a new git tag using released version, i.e. v1.13.1
+# 3. [Create tag] create a new git tag using released version, i.e. ApiGatewayV1.13.1
```
Comment (sdangol) [inline .github/workflows/make-release.yml]: Probably an unintended change

Comment (svozza) [reply .github/workflows/make-release.yml]: Yeah, find and replace went wrong. 😓 Fixed now.
